client/v2/client.go

```diff
@@ -0,0 +1,371 @@
+
+	// InsecureSkipVerify gets passed to the http client, if true, it will
+	// skip https certificate verification. Defaults to false
+	InsecureSkipVerify bool
+}
+
+type BatchPointsConfig struct {
+	// Precision is the write precision of the points, defaults to "ns"
+	Precision string
+
+	// Database is the database to write points to
+	Database string
+
+	// RetentionPolicy is the retention policy of the points
+	RetentionPolicy string
+
+	// Write consistency is the number of servers required to confirm write
+	WriteConsistency string
+}
+
+type Client interface {
+	// Write takes a BatchPoints object and writes all Points to InfluxDB.
+	Write(bp BatchPoints) error
+
+	// Query makes an InfluxDB Query on the database
+	Query(q Query) (*Response, error)
+}
+
+// NewClient creates a client interface from the given config.
+func NewClient(conf Config) Client {
+	if conf.UserAgent == "" {
+		conf.UserAgent = "InfluxDBClient"
+	}
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{
+			InsecureSkipVerify: conf.InsecureSkipVerify,
+		},
+	}
+	return &client{
+		url:       conf.URL,
+		username:  conf.Username,
+		password:  conf.Password,
+		useragent: conf.UserAgent,
+		httpClient: &http.Client{
+			Timeout:   conf.Timeout,
+			Transport: tr,
+		},
+	}
+}
+
+type client struct {
+	url        *url.URL
+	username   string
+	password   string
+	useragent  string
+	httpClient *http.Client
+}
+
+// BatchPoints is an interface into a batched grouping of points to write into
+// InfluxDB together.
+type BatchPoints interface {
+	// AddPoint adds the given point to the Batch of points
+	AddPoint(p *Point)
+	// Points lists the points in the Batch
+	Points() []*Point
+
+	// Precision returns the currently set precision of this Batch
+	Precision() string
+	// SetPrecision sets the precision of this batch.
+	SetPrecision(s string)
+
+	// Database returns the currently set database of this Batch
+	Database() string
+	// SetDatabase sets the database of this Batch
+	SetDatabase(s string)
+
+	// WriteConsistency returns the currently set write consistency of this Batch
+	WriteConsistency() string
+	// SetWriteConsistency sets the write consistency of this Batch
+	SetWriteConsistency(s string)
+
+	// RetentionPolicy returns the currently set retention policy of this Batch
+	RetentionPolicy() string
+	// SetRetentionPolicy sets the retention policy of this Batch
+	SetRetentionPolicy(s string)
+}
+
+// NewBatchPoints returns a BatchPoints interface based on the given config.
+func NewBatchPoints(c BatchPointsConfig) BatchPoints {
+	return &batchpoints{
+		database:         c.Database,
+		precision:        c.Precision,
+		retentionPolicy:  c.RetentionPolicy,
+		writeConsistency: c.WriteConsistency,
+	}
+}
+
+type batchpoints struct {
+	sync.Mutex
+
+	points           []*Point
+	database         string
+	precision        string
+	retentionPolicy  string
+	writeConsistency string
+}
+
+func (bp *batchpoints) AddPoint(p *Point) {
+	bp.Lock()
```

Comment:
In Telegraf the use-case is that we have a batch of points that we write into multiple outputs (each in it's own thread). Some other users may want to write the same points to multiple servers, databases, or retention policies. This would allow them to multi-thread that without having to deepcopy the entire batch.